layout/plan.h

```diff
@@ -0,0 +1,619 @@
+ * 		compute [label="erasure(buf1, buf2, buf-parity)"];
+ * 		readout0p [label="buf0.ready"];
+ * 		compute -> readparity;
+ * 		compute -> read1;
+ * 		compute -> read2;
+ * 		readout0p -> compute;
+ * 	}
+ * 	subgraph cluster_2 {
+ * 		style=filled;
+ * 		color=lightgrey;
+ * 		done
+ * 		done -> readout0p [style=invis];
+ * 		done -> readout1 [style=invis];
+ * 		done -> readout2 [style=invis];
+ * 	}
+ * }
+ * @enddot
+ *
+ * ISC use case
+ * ------------
+ *
+ * ISC user should analyse the plops and their dependencies in order to
+ * figure out whether the user data can be obtained at the server side so
+ * that the computation could be delegated there. In some cases, however,
+ * the data may not be accessible at the server. For example, in degraded
+ * mode some data units in some parity groups have to be recovered from
+ * other data and parity units which may be located on different servers.
+ * In this case the computation on the recovered units is better to do at
+ * the client side.
+ *
+ * If the graph dependency looks like a linear chain starting from M0_LAT_READ
+ * and finishing with M0_LAT_OUT_READ, this might be a good canditate for ISC:
+ *
+ * @verbatim
+ *     M0_LAT_READ <-- [ M0_LAT_FUN <-- ... ] <-- M0_LAT_OUT_READ
+ * @endverbatim
+ *
+ * it means that the user data can be obtained at the server side
+ * independently from the other servers (i.e. there are no several M0_LAT_READ
+ * or parity calculation (M0_LAT_FUN) plops in the dependencies) and the
+ * execution of such a chain of plops can be delegated to the server side
+ * along with the ISC computation.
+ *
+ * ISC code should only make sure that the functions in between the chain
+ * can be run at the server side.
+ *
+ * In all other cases the user data should be retrieved at the client side
+ * and the ISC computation on it should be done at the client side as well.
+ *
+ * @{
+ */
+
+#include "lib/types.h"        /* uint64_t */
+#include "lib/tlist.h"
+#include "lib/vec.h"
+
+struct m0_layout;
+struct m0_layout_plan;
+struct m0_layout_plop;
+struct m0_layout_plop_ops;
+
+/**
+ * Types of plops.
+ *
+ * A value of this enum is placed by the implementation in
+ * m0_layout_plop::pl_type field of a plop, before it is returned by
+ * m0_layout_plan_get().
+ */
+enum m0_layout_plop_type {
+	/**
+	 * Input plop, m0_layout_inout_plop.
+	 *
+	 * Input plop has no dependencies. It is used to signal the
+	 * implementation that a part of input data is available.
+	 *
+	 * An input plop is ready immediately after the plan for a write
+	 * operation is created. The user handles it by setting
+	 * m0_layout_in_object_plop::iob_ready to the set of available data
+	 * extents. More input plops are generated by the implementation until
+	 * the entire range of data is available.
+	 *
+	 * This plop is useful in situations when the entire input data-set is
+	 * not available at once.
+	 *
+	 * @see m0_layout_inout_plop
+	 */
+	M0_LAT_IN_WRITE,
```

Comment:
s/M0_LAT_IN_WRITE/M0_LPT_IN_WRITE
?

similar to M0_LPS_INIT.

Reply:
Fine with me, but it's the original naming from @nikitadanilov. We can update if it's fine with him.